src/Uno.UI/UI/Xaml/UIElement.Layout.netstd.cs

```diff
@@ -4,98 +4,204 @@
 using System.Diagnostics;
 using Windows.UI.Xaml.Controls.Primitives;
 using System.Runtime.CompilerServices;
+using Uno.UI;
 
 namespace Windows.UI.Xaml
 {
 	public partial class UIElement : DependencyObject
 	{
 		private Size _size;
 
-		private bool _isMeasureValid = false;
-		private bool _isArrangeValid = false;
-
 		public Size DesiredSize => Visibility == Visibility.Collapsed ? new Size(0, 0) : ((IUIElement)this).DesiredSize;
 
-		internal bool IsMeasureDirty => !_isMeasureValid;
-		internal bool IsArrangeDirty => !_isArrangeValid;
-
 		/// <summary>
-		/// When set, measure and invalidate requests will not be propagated further up the visual tree, ie they won't trigger a relayout.
+		/// When set, measure and invalidate requests will not be propagated further up the visual tree, ie they won't trigger a re-layout.
 		/// Used where repeated unnecessary measure/arrange passes would be unacceptable for performance (eg scrolling in a list).
 		/// </summary>
 		internal bool ShouldInterceptInvalidate { get; set; }
 
+		[Flags]
+		private protected enum LayoutFlag : byte
+		{
+			/// <summary>
+			/// Means the Measure is dirty for the current element
+			/// </summary>
+			MeasureDirty = 0b0000_0001,
+
+			/// <summary>
+			/// Means the Measure is dirty on at least one child of this element
+			/// </summary>
+			MeasureDirtyPath = 0b0000_0010,
+
+			/// <summary>
+			/// Indicated the first measure has been done on the element after been connected to parent
+			/// </summary>
+			FirstMeasureDone = 0b0000_0100,
+
+			/// <summary>
+			/// Means the Arrange is dirty on the current element or one of its child
+			/// </summary>
+			ArrangeDirty = 0b0001_0000,
+
+			// ArrangeDirtyPath not implemented yet
+		}
+
+		private const LayoutFlag DEFAULT_STARTING_LAYOUTFLAGS = 0;
+
+		private LayoutFlag _layoutFlags = DEFAULT_STARTING_LAYOUTFLAGS;
+
+		/// <summary>
+		/// Check for one specific layout flag
+		/// </summary>
+		[MethodImpl(MethodImplOptions.AggressiveInlining)]
+		private protected bool GetLayoutFlag(LayoutFlag flag) => (_layoutFlags & flag) == flag;
```

Comment:
```suggestion
		private protected bool IsLayoutFlagSet(LayoutFlag flag) => (_layoutFlags & flag) == flag;
```

Reply:
Actually we could also name it `IsFlagSet` or even just `IsSet` as we will have other enum types for other partial classes ^^